src/tools/r2rdump/R2RReader.cs

```diff
@@ -111,12 +110,33 @@ public unsafe R2RReader(string filename)
                         runtimeFunctions[i] = new RuntimeFunction(nativeCodeStartRva, nativeCodeEndRva, nativeCodeUnwindRva);
                     }
 
-                    i = 0;
                     int methodDefEntryPointsRVA = R2RHeader.Sections[R2RSection.SectionType.READYTORUN_SECTION_METHODDEF_ENTRYPOINTS].RelativeVirtualAddress;
                     int methodDefEntryPointsOffset = GetOffset(methodDefEntryPointsRVA);
+                    Dictionary<int, MethodDefinitionHandle> ridToMethodDefHandle = new Dictionary<int, MethodDefinitionHandle>();
                     foreach (MethodDefinitionHandle methodDefHandle in mdReader.MethodDefinitions)
                     {
-                        R2RMethods[i++] = new R2RMethod(_image, runtimeFunctions, ref mdReader, methodDefHandle, methodDefEntryPointsOffset);
+                        ridToMethodDefHandle[MetadataTokens.GetRowNumber(methodDefHandle)] = methodDefHandle;
+                    }
+
+                    i = 0;
+                    NativeArray methodEntryPoints = new NativeArray(_image, (uint)methodDefEntryPointsOffset);
+                    uint nMethods = methodEntryPoints.GetCount();
+                    R2RMethods = new R2RMethod[nMethods];
+                    for (int rid = 1; rid <= nMethods; rid++)
+                    {
+                        uint offset = 0;
+                        if (methodEntryPoints.TryGetAt(_image, rid - 1, ref offset))
+                        {
+                            MethodDefinitionHandle methodDefHandle;
+                            if (ridToMethodDefHandle.TryGetValue(rid, out methodDefHandle))
+                            {
+                                R2RMethods[i++] = new R2RMethod(_image, runtimeFunctions, ref mdReader, methodDefHandle, methodEntryPoints, offset, rid);
+                            }
+                            else
+                            {
+                                R2RMethods[i++] = new R2RMethod(_image, runtimeFunctions, methodEntryPoints, offset, rid);
```

Comment:
Did you encounter a scenario where you hit this else statement? In other words, when there was no method def?

Reply:
I didn't encounter such a scenario, but I've only tested with a simple program. What kind of function might cause it to hit the else statement?

Reply:
I don't know, I suspect we should always have a method definition handle. I don't think this requires any changes right now.